src/main/java/org/opensearch/neuralsearch/processor/normalization/MinMaxScoreNormalizationTechnique.java

```diff
@@ -167,22 +224,197 @@ private float[] getMinScores(final List<CompoundTopDocs> queryTopDocs, final int
+                "min_score must be a valid finite number between %f and %f",
+                LowerBound.MIN_LOWER_BOUND_SCORE,
+                LowerBound.MAX_LOWER_BOUND_SCORE
+            );
+            return minScore;
+        } catch (NumberFormatException e) {
+            throw new IllegalArgumentException("invalid format for min_score: must be a valid float value", e);
+        }
     }
 
     /**
      * Result class to hold min and max scores for each sub query
      */
     @AllArgsConstructor
     @Getter
-    private class MinMaxScores {
+    private static class MinMaxScores {
         float[] minScoresPerSubquery;
         float[] maxScoresPerSubquery;
     }
+
+    /**
+     * Result class to hold lower bound for each sub query
+     */
+    @Getter
+    public static class LowerBound {
+        static final float MIN_LOWER_BOUND_SCORE = -10_000f;
+        static final float MAX_LOWER_BOUND_SCORE = 10_000f;
+        static final float DEFAULT_LOWER_BOUND_SCORE = 0.0f;
+
+        private final boolean enabled;
+        private final Mode mode;
+        private final float minScore;
+
+        LowerBound() {
+            this(false, Mode.DEFAULT, DEFAULT_LOWER_BOUND_SCORE);
+        }
+
+        LowerBound(boolean enabled, Mode mode, float minScore) {
+            this.enabled = enabled;
+            this.mode = mode;
+            this.minScore = minScore;
+        }
+
+        /**
+         * Enum for normalization mode
+         */
+        protected enum Mode {
+            APPLY {
+                @Override
+                public float normalize(float score, float minScore, float maxScore, float lowerBoundScore) {
+                    // if we apply the lower bound this mean we use actual score in case it's less then the lower bound min score
+                    // same applied to case when actual max_score is less than lower bound min score
+                    if (maxScore < lowerBoundScore || score < lowerBoundScore) {
+                        return (score - minScore) / (maxScore - minScore);
+                    }
+                    return (score - lowerBoundScore) / (maxScore - lowerBoundScore);
+                }
+            },
+            CLIP {
+                @Override
+                public float normalize(float score, float minScore, float maxScore, float lowerBoundScore) {
+                    // apply clipping, return lower bound min score if score is less than min score. This effectively means 0 after
+                    // normalization
+                    if (score < minScore) {
+                        return 0.0f;
+                    }
+                    if (maxScore < lowerBoundScore) {
+                        return (score - minScore) / (maxScore - minScore);
+                    }
+                    return (score - lowerBoundScore) / (maxScore - lowerBoundScore);
+                }
+            },
+            IGNORE {
```

Comment:
Do we need a separate enum for the original logic for min max? When the mode is IGNORE we could rely on the original normalize method of MinMax class

Reply:
we could for sure, but as you mentioned we're adding "ignore" as one of the new modes. What I can do is to refactor this logic into a common method that is used from both places: when not lower bound provided, and for "ignore" mode

Reply:
we already going just that, it no lower bounds we do call Mode.IGNORE.normalize() https://github.com/martin-gaievski/neural-search/blob/feature/lower_bounds_for_minmax_normalization/src/main/java/org/opensearch/neuralsearch/processor/normalization/MinMaxScoreNormalizationTechnique.java#L232-L233